pkg/protocol/stan/protocol.go

```diff
@@ -0,0 +1,106 @@
+package stan
+
+import (
+	"context"
+	"errors"
+	"github.com/cloudevents/sdk-go/pkg/binding"
+	"github.com/nats-io/stan.go"
+)
+
+var ErrSubscriptionAlreadyOpen = errors.New("subscription already open")
+
+// Protocol is a reference implementation for using the CloudEvents binding
+// integration. Protocol acts as both a STAN client and a STAN handler.
+type Protocol struct {
+	Conn    stan.Conn
+	Subject string
+
+	Consumer        *Consumer
+	consumerOptions []ConsumerOption
+
+	Sender        *Sender
+	senderOptions []SenderOption
+
+	connOwned bool // whether this protocol created the stan connection
+}
+
+// New creates a new STAN protocol including managing the lifecycle of the connection
+func NewProtocol(clusterID, clientID, subject string, stanOpts []stan.Option, opts ...ProtocolOption) (*Protocol, error) {
+	conn, err := stan.Connect(clusterID, clientID, stanOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := NewProtocolFromConn(conn, subject, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	p.connOwned = true
+
+	return p, nil
+}
+
+// NewProtocolFromConn creates a new STAN protocol but leaves managing the lifecycle of the connection up to the caller
+func NewProtocolFromConn(conn stan.Conn, subject string, opts ...ProtocolOption) (*Protocol, error) {
+	var err error
+	p := &Protocol{
+		Conn:    conn,
+		Subject: subject,
+	}
+
+	if err := p.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	if p.Consumer, err = NewConsumerFromConn(conn, subject, p.consumerOptions...); err != nil {
+		return nil, err
+	}
+
+	if p.Sender, err = NewSenderFromConn(conn, subject, p.senderOptions...); err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+func (p *Protocol) applyOptions(opts ...ProtocolOption) error {
+	for _, fn := range opts {
+		if err := fn(p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Send implements Sender.Send
+func (p *Protocol) Send(ctx context.Context, in binding.Message) error {
+	return p.Sender.Send(ctx, in)
+}
+
+// OpenInbound implements Opener.OpenInbound
+func (p *Protocol) OpenInbound(ctx context.Context) error {
+	return p.Consumer.OpenInbound(ctx)
```

Comment:
@n3wscott @slinkydeveloper regarding the slack conversation around how the `client` package handles closing (or lack of). If we relied on closing the context in `client.StartReceiver` here only p.Consumer would be closed, and the connection would remain open. 

Really the `client` package needs to call `protocol.Close()` which can handle all of the connection, sender and consumer.